### PR TITLE
Fixes for loading dependencies

### LINKED ISSUE #41982: The dictionary the table depends on breaks the clickhouse when drop it
When you try to drop a dictionary on which the table depends, the clickhouse thinks that the dictionary exists and does not exist at the same time. This bug is reproduced on any version ( 22.1.3.7 , 22.3.11.12 , 22.9.3). It is treated by restart

**How to reproduce**

```
CREATE VIEW default.view AS SELECT toUInt32(1) as id, 'v' as value;

CREATE DICTIONARY default.dict (id UInt32, value String) 
PRIMARY KEY id 
SOURCE(CLICKHOUSE(host 'localhost' port 9000 user 'default' db 'default' table 'view')) 
LAYOUT (HASHED()) LIFETIME (600);

CREATE TABLE default.table
(
  col String MATERIALIZED dictGet('default.dict', 'value', toUInt32(1))
) 
ENGINE = MergeTree() 
ORDER BY tuple();

DROP DICTIONARY default.dict;

SELECT * FROM dictionary('default.dict');

SYSTEM RELOAD DICTIONARY default.dict;


CREATE DICTIONARY default.dict (id UInt32, value String) 
PRIMARY KEY id 
SOURCE(CLICKHOUSE(host 'localhost' port 9000 user 'default' db 'default' table 'view')) 
LAYOUT (HASHED()) LIFETIME (600);
```

**Errors**

This error is OK
```
DROP DICTIONARY default.dict

Received exception from server (version 22.9.3):
Code: 630. DB::Exception: Received from clickhouse-server:9000. DB::Exception: Cannot drop or rename default.dict, because some tables depend on it: default.table. (HAVE_DEPENDENT_OBJECTS)
```

And the following is not OK
```
SELECT *
FROM dictionary('default.dict')

Received exception from server (version 22.9.3):
Code: 36. DB::Exception: Received from clickhouse-server:9000. DB::Exception: Dictionary (`default.dict`) not found. (BAD_ARGUMENTS)

---

SYSTEM RELOAD DICTIONARY default.dict

Received exception from server (version 22.9.3):
Code: 36. DB::Exception: Received from clickhouse-server:9000. DB::Exception: Dictionary (`default.dict`) not found. (BAD_ARGUMENTS)
```

```
CREATE DICTIONARY default.dict
(
    `id` UInt32,
    `value` String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(HOST 'localhost' PORT 9000 USER 'default' DB 'default' TABLE 'view'))
LIFETIME(MIN 0 MAX 600)
LAYOUT(HASHED())

Received exception from server (version 22.9.3):
Code: 387. DB::Exception: Received from clickhouse-server:9000. DB::Exception: Dictionary default.dict already exists. (DICTIONARY_ALREADY_EXISTS)
```

### PR DESCRIPTION
**Pull request by tavplubix**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
A table might be shut down and a dictionary might be detached before checking if can be dropped without breaking dependencies between table, it's fixed. Fixes https://github.com/ClickHouse/ClickHouse/issues/41982

### CODE REVIEW

**Comment by m1khal3v** [comment]
Is this issue fixing this too? #41327

**Comment by tavplubix** [comment]
No, it does not

File: src/Databases/DDLDependencyVisitor.cpp [hunk 1]
```diff
@@ -140,4 +149,50 @@ void DDLDependencyVisitor::extractTableNameFromArgument(const ASTFunction & func
...
+void NormalizeAndEvaluateConstants::visit(const ASTFunctionWithKeyValueArguments & dict_source, Data & data)
+{
+    if (!dict_source.elements)
+        return;
+
+    auto & expr_list = dict_source.elements->as<ASTExpressionList &>();
+    for (auto & child : expr_list.children)
+    {
+        ASTPair * pair = child->as<ASTPair>();
+        if (const auto * func = pair->second->as<ASTFunction>())
```
Comment (SmitaRKulkarni) [inline src/Databases/DDLDependencyVisitor.cpp:190]:
```suggestion
        if (pair->second->as<ASTFunction>())
```

File: src/Interpreters/DatabaseCatalog.cpp [hunk 2]
```diff
@@ -1103,16 +1103,7 @@ TableNamesSet DatabaseCatalog::tryRemoveLoadingDependenciesUnlocked(const Qualif
...
-            TableNames from_other_databases;
-            for (const auto & table : dependent)
-                if (table.database != removing_table.database)
-                    from_other_databases.push_back(table);
-
-            if (!from_other_databases.empty())
-                throw Exception(ErrorCodes::HAVE_DEPENDENT_OBJECTS, "Cannot drop or rename {}, because some tables depend on it: {}",
-                                removing_table, fmt::join(from_other_databases, ", "));
-        }
+            checkTableCanBeRemovedOrRenamedImpl(dependent, removing_table, is_drop_database);
```
Comment (SmitaRKulkarni) [inline src/Interpreters/DatabaseCatalog.cpp:1106]: I think we can remove the above checks i.e. `if (check_dependencies && !is_drop_database)` and `!dependent.empty()` as `checkTableCanBeRemovedOrRenamedImpl` will take care of it.

File: src/Interpreters/DatabaseCatalog.h [hunk 3]
```diff
@@ -245,6 +245,7 @@ class DatabaseCatalog : boost::noncopyable, WithMutableContext
 
     void shutdownImpl();
 
+    static void checkTableCanBeRemovedOrRenamedImpl(const TableNamesSet & dependent, const QualifiedTableName & removing_table, bool is_drop_database);
```
Comment (SmitaRKulkarni) [inline src/Interpreters/DatabaseCatalog.h:248]: How about calling this `checkTableDependencies` ?  as we call it even outside of `checkTableCanBeRemovedOrRenamed`

File: src/Interpreters/DatabaseCatalog.cpp (hunk 2 above)
Comment (tavplubix) [reply src/Interpreters/DatabaseCatalog.cpp:1106]: Yep, I forgot to remove it, thanks

File: src/Interpreters/DatabaseCatalog.h (hunk 3 above)
Comment (tavplubix) [reply src/Interpreters/DatabaseCatalog.h:248]: But it's not just "checking dependencies" (whatever it means), it checks if table can be removed or renamed. So I think that `checkTableCanBeRemovedOrRenamedImpl` is more clear

**Review by SmitaRKulkarni — APPROVED** [review]
LGTM

**Comment by tavplubix** [comment]
Stateless tests (aarch64) - 02267_file_globs_schema_inference is still broken
Stress test (asan) - https://github.com/ClickHouse/ClickHouse/issues/39174
Stress test (debug) - https://github.com/ClickHouse/ClickHouse/issues/39174
Stress test (msan) - https://github.com/ClickHouse/ClickHouse/issues/41831
Stress test (tsan) - `time="2022-10-10T16:41:54Z" level=error msg="error waiting for container: unexpected EOF"` - some issue with infrastructure 
Stress test (ubsan) - looks like https://github.com/ClickHouse/ClickHouse/issues/41409

**Pull request merged**